experimental/packages/exporter-trace-otlp-http/test/node/CollectorTraceExporter.test.ts

```diff
@@ -408,6 +408,45 @@ describe('OTLPTraceExporter - node with json over http', () => {
     });
   });
 
+  describe('export - with dynamic headers', () => {
+    let count = 0;
+    beforeEach(() => {
+      collectorExporterConfig = {
+        headers: () => ({
+          foo: `dynamic-${count}`,
+        }),
+        hostname: 'foo',
+        url: 'http://foo.bar.com',
+        keepAlive: true,
+        httpAgentOptions: { keepAliveMsecs: 2000 },
+      };
+      collectorExporter = new OTLPTraceExporter(collectorExporterConfig);
+      spans = [];
+      spans.push(Object.assign({}, mockedReadableSpan));
+    });
+
+    it('should set dynamic custom headers', done => {
+      setTimeout(() => {
+        for (let i = 2; i < 5; i++) {
+          fakeRequest = new Stream.PassThrough();
+          sinon.restore();
```

Comment:
not sure if this was copied from another test case, bu should this be in a before/after function?

Reply:
What I was trying to do here was actually test that the headers are dynamically applied. In particular the response of the function is not cached (which was a bug an early version of this change had). This test is looping over a sequence of calls and just using the count as a way to check the expected dynamic header is there. In the real world example this would be the bearer token that would be updated when needed.

The `sinon.restore()` in the loop is to allow `http.request` to be re-stubbed and hence why it is within the loop not in a before / after function.  Although normally we'd see it in clean up code before or after test (for guaranteed clean up) I suspect it can be dropped in anywhere it is needed.

Happy to consider alternative test approach here. What I want is to verify is that the implementation does allow dynamic changes and doesn't inadvertently cache the response from the function.

Reply:
Ah gotcha. My concern was mostly that it may accidentally restore stubs from a higher scope `beforeEach()`, since it is a global operation. I think it doesn't matter now but the test would behave differently if more stubbing is added in the future

<!--
I think you can reset just the single stub to future proof the test if more stubs are added

```ts
stubRequest.reset();
stubRequest.returns(fakeRequest as any);
```

Or alternatively use callsFake to always supply a new request object
-->